lib/codecs/src/decoding/format/gelf.rs

```diff
@@ -0,0 +1,396 @@
+    pub fn new() -> GelfDeserializer {
+        GelfDeserializer
+    }
+
+    /// Builds a LogEvent from the parsed GelfMessage.
+    /// The logic follows strictly the documented GELF standard.
+    fn message_to_event(&self, parsed: &GelfMessage) -> vector_core::Result<Event> {
+        let mut log = LogEvent::from(parsed.short_message.to_string());
+
+        // GELF spec defines the version as 1.1 which has not changed since 2013
+        if parsed.version != GELF_VERSION {
+            return Err(format!(
+                "{} does not match GELF spec version ({})",
+                VERSION, GELF_VERSION
+            )
+            .into());
+        }
+
+        log.insert(VERSION, parsed.version.to_string());
+        log.insert(HOST, parsed.host.to_string());
+
+        if let Some(full_message) = &parsed.full_message {
+            log.insert(FULL_MESSAGE, full_message.to_string());
+        }
+
+        if let Some(timestamp) = parsed.timestamp {
+            let naive = NaiveDateTime::from_timestamp(
+                f64::trunc(timestamp) as i64,
+                f64::fract(timestamp) as u32,
+            );
+            log.insert(
+                log_schema().timestamp_key(),
+                DateTime::<Utc>::from_utc(naive, Utc),
+            );
+        // per GELF spec- add timestamp if not provided
+        } else {
+            log.insert(log_schema().timestamp_key(), Utc::now());
+        }
+
+        if let Some(level) = parsed.level {
+            log.insert(LEVEL, level);
+        }
+        if let Some(facility) = &parsed.facility {
+            log.insert(FACILITY, facility.to_string());
+        }
+        if let Some(line) = parsed.line {
+            log.insert(
+                LINE,
+                value::Value::Float(ordered_float::NotNan::new(line).unwrap()),
+            );
+        }
+        if let Some(file) = &parsed.file {
+            log.insert(FILE, file.to_string());
+        }
+
+        if let Some(add) = &parsed.additional_fields {
+            for (key, val) in add.iter() {
+                // per GELF spec, filter out _id
+                if key == "_id" {
+                    continue;
+                }
+                // per GELF spec, Additional field names must be characters dashes or dots
+                if !VALID_FIELD_REGEX.is_match(key) {
+                    if key.starts_with('_') {
+                        return Err(format!("'{}' field is invalid. \
+                                           Additional field names must be prefixed with an underscore.", key).into());
+                    } else {
+                        return Err(format!("'{}' field contains invalid characters. Field names may \
+                                           contain only letters, numbers, underscores, dashes and dots.", key).into());
+                    }
+                }
+
+                let vector_val: value::Value = val.into();
+                log.insert(path!(key.as_str()), vector_val);
```

Comment:
Technically the spec, https://docs.graylog.org/docs/gelf, says that additional fields can only be strings or numbers. This allows for other JSON values like booleans, objects, arrays, and `null`. I guess, in the spirit of being strict with the option to relax later, we can check for that with `match` over the `serde_json::Value`.

Reply:
I'm not seeing a way to use match over `serde_json::Value` , there are the functions to check if it is a specific type however, which should work.